src/NumericalAlgorithms/FiniteDifference/Reconstruct.hpp

```diff
@@ -0,0 +1,70 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <utility>  // for std::pair
+
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+template <size_t Dim>
+class Direction;
+template <size_t Dim, typename T>
+class DirectionMap;
+template <size_t Dim>
+class Index;
+/// \endcond
+
+namespace fd {
+/*!
+ * \ingroup FiniteDifferenceGroup
+ * \brief Variable and flux vector splitting reconstruction schemes for finite
+ * difference methods.
+ *
+ * Implementations of reconstruction methods must call this function
+ * (`fd::reconstruction::detail::reconstruct`) to perform the reconstruction.
+ * This function performs the actual loops taking into account strides and ghost
+ * zones for the reconstruction method. The type of reconstruction done is
+ * specified with the `Reconstructor` explicit template parameter. Parameters
+ * for the reconstruction scheme are forwarded along using the
+ * `args_for_reconstruction` parameter pack.
+ *
+ * `Reconstructor` classes must define:
+ * - a `static constexpr size_t stencil_width()` function that
+ *   returns the size of the stencil, e.g. 2 for minmod, and 5 for 5-point
+ *   reconstruction.
+ * - a
+ *   \code
+ *      SPECTRE_ALWAYS_INLINE static std::array<double, 2> pointwise(
+ *            const double* const q, const int stride)
+ *   \endcode
+ *   function that optionally takes the additional arguments.
+ *   The `q` are the input values to reconstruct. The value \f$q_i\f$ in the
+ *   current FD cell is located at `q[0]`, the value at neighbor cell
+ *   \f$q_{i+1}\f$ is at `q[stride]`, and the value at neighbor cell
+ *   \f$q_{i-1}\f$ is at `q[-stride]`. The returned values are the
+ *   reconstructed solution on the lower and upper side of the cell.
+ *
+ * \note Currently the stride is always one because we transpose the data before
+ * reconstruction. However, it may be faster to have a non-unit stride without
+ * the transpose. We have the `stride` parameter in the reconstruction schemes
+ * to make testing performance easier in the future.
+ */
+namespace reconstruction {
```

Comment:
I think this documentation could be improved with a figure illustrating cell-centered points, face-centered points and what the reconstruct function is returning; I found myself getting confused by the upper_side_of_face being filled with the lower side of the cell returned by the pointwise function.  I don't want this to hold up the PR, so you can defer it as an issue if you want.